### PR TITLE
feat: collect coverage of end-to-end tests

### PR DESCRIPTION
**Pull request by mxschmitt**
TODOs:
- [x] add comment for documentation in global d.ts what `saveCoverage` does
- [x] add documentation in README: required following babel plugin: https://www.npmjs.com/package/babel-plugin-istanbul
- [x] clarify with @mmarkelov if loading the config like I did in `PlaywrightRunner.runTests` is okay

Example usage here: https://github.com/playwright-community/playwright-jest-examples/pull/8

Will look like that if you are using the example: 
image

### CODE REVIEW

**Comment by mmarkelov** [comment]
@mxschmitt that's a really nice new feature. But maybe we should put it in separate package and make it optional?

**Comment by mxschmitt** [comment]
> @mxschmitt that's a really nice new feature. But maybe we should put it in separate package and make it optional?

Unfortunately it needs a tight integrations in before all tests and after all tests hooks. That's why I think it does not make so much sense and would decrease the user experience for that significantly. As a pitfall it adds a few more dependencies, but shouldn't be a big deal in my mind.

File: src/PlaywrightRunner.ts [hunk 1]
```diff
@@ -120,8 +125,11 @@ class PlaywrightRunner extends JestRunner {
     onFailure: OnTestFailure,
     options: TestRunnerOptions,
   ): Promise<void> {
-    const browserTests = await this.getTests(tests)
-
+    const config = await readConfig(tests[0].context.config.rootDir)
```
Comment (mmarkelov) [inline src/PlaywrightRunner.ts:128]: I'm not sure that using `rootDir` from first test is a good decision

File: src/PlaywrightEnvironment.ts [hunk 2]
```diff
@@ -140,6 +141,8 @@ export const getPlaywrightEnv = (basicEnv = 'node') => {
             stdin.on('data', onKeyPress)
           })
         },
+        saveCoverage: async (page: Page): Promise<void> =>
+          saveCoverage(page, config.collectCoverage),
```
Comment (mmarkelov) [inline src/PlaywrightEnvironment.ts:145]: A little confused about naming. Maybe we should import it with another name?

File: src/PlaywrightRunner.ts (hunk 1 above)
Comment (mxschmitt) [reply src/PlaywrightRunner.ts:128]: We could also use the global rootDir, do you think this is better? Also not sure tho

Comment (mmarkelov) [reply src/PlaywrightRunner.ts:128]: > We could also use the global rootDir, do you think this is better? Also not sure tho

global rootDir won't work :( https://github.com/playwright-community/jest-playwright/issues/120

Comment (mxschmitt) [reply src/PlaywrightRunner.ts:128]: Then we have to make it either per test which would be super complicated or let it as it is. Should be probably for most use cases or what do you think?

Comment (mmarkelov) [reply src/PlaywrightRunner.ts:128]: @mxschmitt I think we can test it out for a while. If there will be some issues related to `rootDir` we can fix it

File: src/PlaywrightEnvironment.ts (hunk 2 above)
Comment (mxschmitt) [reply src/PlaywrightEnvironment.ts:145]: Hm, ye. But not sure how to name it.

Comment (mmarkelov) [reply src/PlaywrightEnvironment.ts:145]: > Hm, ye. But not sure how to name it.

IDK, maybe `rootSaveCoverage`?

File: src/PlaywrightRunner.ts (hunk 1 above)
Comment (mxschmitt) [reply src/PlaywrightRunner.ts:128]: so whats better, the global rootDir or the rootDir of the first test?

File: src/PlaywrightEnvironment.ts (hunk 2 above)
Comment (mxschmitt) [reply src/PlaywrightEnvironment.ts]: I choose `savePageCoverage`, should be also okay

**Comment by mxschmitt** [comment]
Let me know if they are any other issues before we can merge it. Thanks! :+1:

File: src/PlaywrightRunner.ts (hunk 1 above)
Comment (mmarkelov) [reply src/PlaywrightRunner.ts:128]: > so whats better, the global rootDir or the rootDir of the first test?

the rootDir of the first test, cause global rootDir is not correct to use for this

**Comment by mmarkelov** [comment]
> Let me know if they are any other issues before we can merge it. Thanks! 👍

I think we are ready to merge it

**Pull request merged**